### DISCUSSION TITLE
Doubts about real impact of bundle size in final percieved UX

### DISCUSSION THREAD

**Discussion by kimgiutae** [vercel/next.js#73956] [Help] [answered]
### Summary

**Hello community!**

_TLDR: Does JS bundle load blocks the render of HTML/CSS in next.js apps ?_

Recently I started to worry about the bundle size of next.js/react.js. Even new versions of both projects don't show a significant reduction of code. Actually, it has increased.

I tried other options like Astro for more content focused projects. But I find myself too used to the “React-way”, which makes me return to the previous Stack (Dunno if it happened to other devs).

I know that in most cases, this shouldn't be a problem. CDN, cache, images optimization and good hosting can tackle most of the issues. Also internet speed is increasing worldwide. At the end of the day, differences can be just milliseconds.

But at least (and as a matter of curiosity), I want to understand the way a next.js v15 app works at a request level.

**When the user accesses a URL in the app, it will load exactly what first?**

As far as I understand, it will load the SSR static version with HTML and CSS of that specific route (layout + page). And the JS should load asynchronously with React doing the “hydration” process later. Am I right?

In that case. **If the JS is not blocking the rendering process (because it is asynchronous), is the bundle size a real problem in modern applications?**

I mean, I could import 100Mb of pointless JS, but theoretically the HTML and CSS of the same route should take the same time to render on screen. It'll take only more time to be interactive, but the user should see the same content at the same time as before.

**I'm wrong?**

**Comment by icyJoseph** [comment]
I mean, Time to Visible, and Time to Interactive, are both very important metrics. I think you need to read up some literature about the critical rendering path:

- https://developer.mozilla.org/en-US/docs/Web/Performance/Critical_rendering_path
- https://web.dev/learn/performance/understanding-the-critical-path

For most apps, the bundle size, unless it is outrageously large, it is more of a business driven requirement. Are you serving people mostly on their phone, on non-wifi connections? 

Tools such as lighthouse would also flag, unused code, and other resources. 

It is also not only the download time that's a problem, but the parsing and evaluation of that code. The slower the device, the worst these get. 

Could you reframe your concern, I am not sure I am following it 100%.

**Reply by kimgiutae** [reply]
Thanks to reply @icyJoseph 

I know that Time to Visible and Time to Interactive are both important.

My question is if bundle size or the way next.js app works affect Time to Visible or that depends exclusively of internet speed and hosting response time.

I mean, two Next.js apps with one 1mb and the other 100mb of bundle sizes, both will have the same Time to Visible speed ?

And also, from the user point of view, im wrong if I supose that TTV is more important than TTI? Maybe the first one gives a better percieved speed UX ?

**Reply by icyJoseph** [reply, marked as answer]
> I mean, two Next.js apps with one 1mb and the other 100mb of bundle sizes, both will have the same Time to Visible speed ?

The larger the HTML you send to the client, the longer the TTV. Frameworks that do hydration, have a problem, that people often call x2 data problem, where the HTML also contains script tags, that contain the input data to be used during hydraiton.

Pages Router used defer, and App Router uses async to load application scripts, so by that nature, those do not block. However, async scripts, might block parsing, if they finish loading before parsing has finished.

- https://pagespeedchecklist.com/async-and-defer
- https://www.kirupa.com/html5/running_your_code_at_the_right_time.htm

The usage of async/defer, and the critical render path, are very important here. Most of the literature you'll find around, points at how important it is to keep resources small, and optimise bundle size.

> And also, from the user point of view, im wrong if I supose that TTV is more important than TTI? Maybe the first one gives a better percieved speed UX ?

TTI trumps. A page that's showing but is unresponsive, is a worse experience.

Perhaps I am lacking some authority here to set the case for TTI, so here's a talk, by a Tech Lead at Wix, from some 5 years ago. Things haven't changed. 

- https://www.youtube.com/watch?v=O1wBZviW7yI&t=452s
- Slides: https://docs.google.com/presentation/d/1D6IzjmjsUKuUkmjzTPg8E32OqQF1MUWMlP4fbgDZH94/edit#slide=id.g57183fade2_0_7

The important point to drive home here:

> - With CSR: Time To Visible and Time To Interactive are one and the same.
> - With SSR: Time To Visible is improve. Time To Interactive at best remains the same.
> 
> Time To Interactive:
> - Not improved by SSR
> - Sometimes worse: CSR → SSR + hydration
> - Unresponsive UI perceived as buggy

In that talk he points to some Akamai research, possibly this one https://speakerdeck.com/bluesmoon/ux-and-performance-metrics-that-matter-a062d37f-e6c7-4b8a-8399-472ec76bb75e, where it is shown that if the TTI is longer than the TTV, "rage clicking" is observed.
